File: src/Atlasd/Battlenet/Protocols/Game/GameKey.cs
﻿using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Linq;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game
{
    class GameKey
    {
        public enum ProductValues : UInt32
        {
            Starcraft_A = 0x01,
            Starcraft_B = 0x02,
            WarcraftII = 0x04,
            DiabloIIBeta = 0x05,
            StarcraftIIBeta = 0x05,
            WorldOfWarcraftWrathOfTheLichKingAlpha = 0x05,
            DiabloII_A = 0x06,
            DiabloII_B = 0x07,
            DiabloIIStressTest = 0x09,
            DiabloIILordOfDestruction_A = 0x0A,
            DiabloIILordOfDestruction_Beta = 0x0B,
            DiabloIILordOfDestruction_B = 0x0C,
            WarcraftIIIBeta = 0x0D,
            WarcraftIIIReignOfChaos_A = 0x0E,
            WarcraftIIIReignOfChaos_B = 0x0F,
            WarcraftIIIFrozenThroneBeta = 0x11,
            WarcraftIIIFrozenThrone_A = 0x12,
            WarcraftIIIFrozenThrone_B = 0x13,
            WorldOfWarcraftBurningCrusade = 0x15,
            WorldOfWarcraft14DayTrial = 0x16,
            Starcraft_DigitalDownload = 0x17,
            DiabloII_DigitalDownload = 0x18,
            DiabloIILordOfDestruction_DigitalDownload = 0x19,
            WorldOfWarcraftWrathOfTheLichKing = 0x1A,
            StarcraftII = 0x1C,
            DiabloIII = 0x1D,
            HeroesOfTheStorm = 0x24,
        };

        public byte[] PrivateValue { get; protected set; }
        public UInt32 ProductValue;
        public UInt32 PublicValue;

        public GameKey(UInt32 keyLength, UInt32 productValue, UInt32 publicValue, byte[] hashedKeyData)
        {
            // This broke joe)x86('s 16 year-old JavaOp2 bot on Fri, Oct 16 2020 :)  -Carl
            if (!(keyLength == 13 || keyLength == 16 || keyLength == 26))
                throw new GameProtocolViolationException(null, "Invalid game key length");

            if (!IsValidProductValue((ProductValues)productValue))
                throw new GameProtocolViolationException(null, "Invalid game key product value");

            SetProductValue(productValue);
            SetPublicValue(publicValue);
            SetPrivateValue(hashedKeyData); // TODO: Do something with hashedKeyData, stuffing it here for now.
        }

        public GameKey(UInt32 productValue, UInt32 publicValue, byte[] privateValue)
        {
            SetPrivateValue(privateValue);
            SetProductValue(productValue);
            SetPublicValue(publicValue);
        }

        public GameKey(string keyString)
        {
            var m_gameKey = new MBNCSUtil.CdKey(keyString);

            if (m_gameKey == null || !m_gameKey.IsValid)
            {
                throw new GameProtocolViolationException(null, "Cannot parse invalid game key");
            }

            SetPrivateValue(m_gameKey.GetValue2());
            SetProductValue((uint)m_gameKey.Product);
            SetPublicValue((uint)m_gameKey.Value1);
        }

        public bool IsValidProductValue()
        {
            return GameKey.IsValidProductValue((ProductValues)ProductValue);
        }

        public static bool IsValidProductValue(ProductValues productValue)
        {
            return productValue switch
            {
                ProductValues.DiabloIIBeta => true,
                ProductValues.DiabloIILordOfDestruction_A => true,
                ProductValues.DiabloIILordOfDestruction_B => true,
                ProductValues.DiabloIILordOfDestruction_Beta => true,
                ProductValues.DiabloIILordOfDestruction_DigitalDownload => true,
                ProductValues.DiabloIIStressTest => true,
                ProductValues.DiabloII_A => true,
                ProductValues.DiabloII_B => true,
                ProductValues.DiabloII_DigitalDownload => true,
                ProductValues.Starcraft_A => true,
                ProductValues.Starcraft_B => true,
                ProductValues.Starcraft_DigitalDownload => true,
                ProductValues.WarcraftII => true,
                ProductValues.WarcraftIIIBeta => true,
                ProductValues.WarcraftIIIFrozenThroneBeta => true,
                ProductValues.WarcraftIIIFrozenThrone_A => true,
                ProductValues.WarcraftIIIFrozenThrone_B => true,
                ProductValues.WarcraftIIIReignOfChaos_A => true,
                ProductValues.WarcraftIIIReignOfChaos_B => true,
                _ => false,
            };
        }

        public static uint RequiredKeyCount(Product.ProductCode code)
        {
            uint count = 0;
            string codeStr = Battlenet.Product.ToString(code);
            string codeStrR = new string(codeStr.Reverse().ToArray());
            if (count == 0) count = Settings.GetUInt32(new string[]{ "battlenet", "emulation", "required_game_key_count", codeStr }, 0, true); // Ltrd
            if (count == 0) count = Settings.GetUInt32(new string[]{ "battlenet", "emulation", "required_game_key_count", codeStrR }, 0, true); // Drtl
            if (count == 0) count = Settings.GetUInt32(new string[]{ "battlenet", "emulation", "required_game_key_count", codeStr.ToUpperInvariant() }, 0, true); // LTRD
            if (count == 0) count = Settings.GetUInt32(new string[]{ "battlenet", "emulation", "required_game_key_count", codeStrR.ToUpperInvariant() }, 0, true); // DRTL
            if (count == 0) count = Settings.GetUInt32(new string[]{ "battlenet", "emulation", "required_game_key_count", codeStr.ToLowerInvariant() }, 0, true); // ltrd
            if (count == 0) count = Settings.GetUInt32(new string[]{ "battlenet", "emulation", "required_game_key_count", codeStrR.ToLowerInvariant() }, 0, true); // drtl
            return count;
        }

        public void SetPrivateValue(byte[] privateValue)
        {
            if (!(privateValue.Length == 4 || privateValue.Length == 20))
                throw new GameProtocolViolationException(null, "Invalid game key private value");

            PrivateValue = privateValue;
        }

        public void SetProductValue(UInt32 productValue)
        {
            ProductValue = productValue;
        }

        public void SetPublicValue(UInt32 publicValue)
        {
            PublicValue = publicValue;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Other/GameKey.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Linq
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game
    Class GameKey
        Public Enum ProductValues As UInt32
            Starcraft_A = &H1
            Starcraft_B = &H2
            WarcraftII = &H4
            DiabloIIBeta = &H5
            StarcraftIIBeta = &H5
            WorldOfWarcraftWrathOfTheLichKingAlpha = &H5
            DiabloII_A = &H6
            DiabloII_B = &H7
            DiabloIIStressTest = &H9
            DiabloIILordOfDestruction_A = &HA
            DiabloIILordOfDestruction_Beta = &HB
            DiabloIILordOfDestruction_B = &HC
            WarcraftIIIBeta = &HD
            WarcraftIIIReignOfChaos_A = &HE
            WarcraftIIIReignOfChaos_B = &HF
            WarcraftIIIFrozenThroneBeta = &H11
            WarcraftIIIFrozenThrone_A = &H12
            WarcraftIIIFrozenThrone_B = &H13
            WorldOfWarcraftBurningCrusade = &H15
            WorldOfWarcraft14DayTrial = &H16
            Starcraft_DigitalDownload = &H17
            DiabloII_DigitalDownload = &H18
            DiabloIILordOfDestruction_DigitalDownload = &H19
            WorldOfWarcraftWrathOfTheLichKing = &H1A
            StarcraftII = &H1C
            DiabloIII = &H1D
            HeroesOfTheStorm = &H24
        End Enum

        Public Property PrivateValue As Byte()
        Public ProductValue As UInt32
        Public PublicValue As UInt32

        Public Sub New(ByVal varKeyLength As UInt32,
                       ByVal varProductValue As UInt32,
                       ByVal varPublicValue As UInt32,
                       ByVal varHashedKeyData As Byte())
            If Not (varKeyLength = 13 OrElse varKeyLength = 16 OrElse varKeyLength = 26) Then Throw New GameProtocolViolationException(Nothing, "Invalid game key length")
            If Not IsValidProductValue(CType(ProductValue, ProductValues)) Then Throw New GameProtocolViolationException(Nothing, "Invalid game key product value")
            SetProductValue(varProductValue)
            SetPublicValue(varPublicValue)
            SetPrivateValue(varHashedKeyData)
        End Sub

        Public Sub New(ByVal varProductValue As UInt32, ByVal varPublicValue As UInt32, ByVal varPrivateValue As Byte())
            SetProductValue(varProductValue)
            SetPublicValue(varPublicValue)
            SetPrivateValue(varPrivateValue)
        End Sub

        Public Sub New(ByVal varKeyString As String)
            Dim m_gameKey = New MBNCSUtil.CdKey(varKeyString)

            If m_gameKey Is Nothing OrElse Not m_gameKey.IsValid Then
                Throw New GameProtocolViolationException(Nothing, "Cannot parse invalid game key")
            End If

            SetPrivateValue(m_gameKey.GetValue2())
            SetProductValue(CUInt(m_gameKey.Product))
            SetPublicValue(CUInt(m_gameKey.Value1))
        End Sub

        Public Function IsValidProductValue() As Boolean
            Return GameKey.IsValidProductValue(CType(ProductValue, ProductValues))
        End Function

        Public Shared Function IsValidProductValue(ByVal varProductValue As ProductValues) As Boolean
            Select Case varProductValue
                Case ProductValues.DiabloIIBeta,
                     ProductValues.DiabloIILordOfDestruction_A,
                     ProductValues.DiabloIILordOfDestruction_B,
                     ProductValues.DiabloIILordOfDestruction_Beta,
                     ProductValues.DiabloIILordOfDestruction_DigitalDownload,
                     ProductValues.DiabloIIStressTest,
                     ProductValues.DiabloII_A,
                     ProductValues.DiabloII_B,
                     ProductValues.DiabloII_DigitalDownload,
                     ProductValues.Starcraft_A,
                     ProductValues.Starcraft_B,
                     ProductValues.Starcraft_DigitalDownload,
                     ProductValues.WarcraftII,
                     ProductValues.WarcraftIIIBeta,
                     ProductValues.WarcraftIIIFrozenThroneBeta,
                     ProductValues.WarcraftIIIFrozenThrone_A,
                     ProductValues.WarcraftIIIFrozenThrone_B,
                     ProductValues.WarcraftIIIReignOfChaos_A,
                     ProductValues.WarcraftIIIReignOfChaos_B
                    Return True
                Case Else
                    Return False
            End Select
        End Function

        Public Shared Function RequiredKeyCount(ByVal varCode As Product.ProductCode) As UInteger
            Dim buf = New Byte(3) {}

            Using m = New MemoryStream(buf)
                Using w = New BinaryWriter(m)
                    w.Write(CUInt(varCode))
                End Using
            End Using

            Dim productStr = Daemon.Common.ReverseString(Encoding.UTF8.GetString(buf))
            Dim battlenetJson = Nothing, emulationJson = Nothing, requiredGameKeyCountJson = Nothing, productJson = Nothing

            Try
                Settings.State.RootElement.TryGetProperty("battlenet", battlenetJson)
                battlenetJson.TryGetProperty("emulation", emulationJson)
                emulationJson.TryGetProperty("required_game_key_count", requiredGameKeyCountJson)
                requiredGameKeyCountJson.TryGetProperty(productStr, productJson)
                Return productJson.GetUInt32()
            Catch ex As Exception
                If Not (TypeOf ex Is ArgumentNullException OrElse TypeOf ex Is InvalidOperationException) Then Throw
                Return 0
            End Try
        End Function

        Public Sub SetPrivateValue(ByVal varPrivateValue As Byte())
            If Not (PrivateValue.Length = 4 OrElse PrivateValue.Length = 20) Then Throw New GameProtocolViolationException(Nothing, "Invalid game key private value")
            PrivateValue = varPrivateValue
        End Sub

        Public Sub SetProductValue(ByVal varProductValue As UInt32)
            ProductValue = varProductValue
        End Sub

        Public Sub SetPublicValue(ByVal varPublicValue As UInt32)
            PublicValue = varPublicValue
        End Sub
    End Class

End Namespace